src/core/utils/Range.hpp

```diff
@@ -0,0 +1,43 @@
+#ifndef UTILS_RANGE_HPP
+#define UTILS_RANGE_HPP
+
+#include <utility>
+
+namespace Utils {
+
+/**
+ * @brief Helper class that wraps two iterators into
+          a range object.
+*/
+template <typename Iterator> class Range {
```

Comment:
To me this looks like a copy of [`boost::iterator_range`](http://www.boost.org/doc/libs/1_63_0/libs/range/doc/html/range/reference/utilities/iterator_range.html).  Why don't you use this?  It seems to be header-only as well.